### ISSUE TITLE
#[var(set = `function_name`)] is not compatible with #[func(rename = `gdscript_name`)]

### ISSUE THREAD

**Issue by Houtamelo** [godot-rust/gdext#863]
Example:
```rust
#[derive(GodotClass)]
#[class(init, base = CharacterBody2D)]
pub struct Herbivore {
	#[var(get, set = gdscript_set_health)]
	pub(super) health: f64,
}

#[godot_api]
impl Herbivore {
	#[func(rename = set_health)]
	fn gdscript_set_health(&mut self, value: f64) { 
		self.set_health(value);
	}
}
```

This compiles in Rust but in Godot we get a runtime error when trying to invoke the property:
![image](https://github.com/user-attachments/assets/22b5f51f-b347-4bc6-a7ce-2763669675b3)

This happens because inside Godot its using the original name, not the renamed one.

Not sure if this is by design or a niche bug.

**Comment by Bromeon** [comment]
Interesting! Certainly not intended.

Might be a bit tricky because those are in two separate macros, and I don't know if the `#[func(rename)]` info is "stored" somewhere to be accessible by code generated from `#[derive(Godot)]`. It _should_ be possible, but might require an extra indirection for each get/set function reference.

**Comment by Bromeon** [comment]
One idea to address this: `#[godot_api]` generates a compile-time lookup table with all `#[func]`s and their registered names. So
```rs
#[godot_api]
impl Herbivore {
	#[func]
	fn eat_herbs(&mut self, value: f64) {...}

	#[func(rename = set_health)]
	fn gdscript_set_health(&mut self, value: f64) {...}
}
```
generates something like
```rs
mod __funcs_Herbivore {
      pub const eat_herbs: &str = "eat_herbs";
      pub const gdscript_set_health: &str = "set_health";
}
```
<br>

Then, the custom setter
```rs
#[var(get, set = gdscript_set_health)]
pub(super) health: f64,
```
would, instead of registering the setter as `"gdscript_set_health"` in Godot, look up the name as `__funcs_Herbivore::gdscript_set_health`. This solves two problems:
1. It takes into account renamed methods.
2. If someone forgets `#[func]` above `gdscript_set_health`, this causes a compile-time error (at the moment: runtime error).

**Comment by 0x53A** [comment]
That got kinda harder with my PR, where you can now have multiple ``#[godot_api]``.

Since you can't contribute to one ``mod`` from multiple places, the constants would have to live in the surrounding module.

```rust
#[godot_api]
impl Herbivore {
	#[func]
	fn eat_herbs(&mut self, value: f64) {...}
}

// generated
pub const __funcs_Herbivore_eat_herbs: &str = "eat_herbs";


#[godot_api]
impl Herbivore {
	#[func(rename = set_health)]
	fn gdscript_set_health(&mut self, value: f64) {...}
}

// generated
pub const __funcs_Herbivore_gdscript_set_health: &str  = "set_health";
```

**Comment by Bromeon** [comment]
Ugh... back to C namespacing then 😔 

(It could be `impl` scoped in this case... but generally I don't like polluting neither the struct's nor the surrounding scope)

**Comment by 0x53A** [comment]
would you merge a PR for this anyway? if yes, what pattern exactly?
```rust
/// Generated by the #[func] macro of gdext. Maps the name under which it is exported to godot to the rust name of the function.
#[doc(hidden)]
pub(self) const __func_{class}_{rust-fn-name}: &str  = "{godot-fn-name}";
```
(I could take a stab at implementing it next week or so)

**Comment by Bromeon** [comment]
Yes, a PR would still be appreciated! It would already fix the issue, and if we find another approach in the future, we can always change it 🙂 

You could maybe use the `__gdext_` prefix like here:
https://github.com/godot-rust/gdext/blob/d32d569ed61ef8427470acdc41e8e708828fe412/godot-core/src/init/mod.rs#L21

Also, [`pub(self)` is redundant](https://stackoverflow.com/questions/50905715/how-is-pubself-visibility-different-from-no-pub-attribute) , so it could be along:
```rs
#[doc(hidden)]
const __gdext_func_{class}_{rust-fn-name}: &str  = "{godot-fn-name}";
```
